metaseq/data/document_to_sequence.py

```diff
@@ -0,0 +1,352 @@
+        self.drop_last = drop_last
+        self.padding_idx = padding_idx
+        self.shuffle_buffer_size = shuffle_buffer_size
+        self.to_skip = to_skip
+        self.permute_documents = permute_documents
+
+        if break_mode == "none":
+            self.block_iterator = yield_token_blocks
+        elif break_mode == "eos_pad_8":
+            self.block_iterator = yield_single_sentences_pad_8
+        elif break_mode == "complete":
+            self.block_iterator = yield_doc_blocks
+        elif break_mode == "passthrough":
+            self.block_iterator = yield_passthrough
+        else:
+            raise ValueError(
+                f'Invalid value for break_mode = {break_mode}. Available options are "none", "eos_pad_8" or "complete".'
+            )
+
+        if not drop_last and padding_idx is None:
+            raise ValueError("padding_idx is required when drop_last is False")
+
+        assert shuffle_buffer_size >= 1
+        if shuffle_buffer_size > 1 and seed is None:
+            raise ValueError("seed is required when shuffle_buffer_size > 1")
+
+        # if break_mode != "none": raise NotImplementedError
+
+        self._started_iteration = False
+
+    def set_epoch(self, epoch):
+        if hasattr(self.dataset, "set_epoch"):
+            self.dataset.set_epoch(epoch)
+
+        if self.permute_documents:
+            # shuffle the dataset according to the seed argument and epoch
+            seed = int(hash((self.document_shuffle_seed, epoch)) % 1e6)
+            with data_utils.numpy_seed(seed):
+                self.indices = np.random.permutation(len(self.dataset))
+        else:
+            self.indices = list(range(len(self.dataset)))
+
+    def set_shuffle_buffer_size(self, new_shuffle_buffer_size):
+        assert not self._started_iteration
+        self.shuffle_buffer_size = new_shuffle_buffer_size
+
+    def __iter__(self):
+        skip_time = 0
+        t0 = time.time()
+        assert not self._started_iteration
+        self.started_iteration = True
+
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is not None and worker_info.num_workers > 1:
+            chunks = np.array_split(self.indices, worker_info.num_workers)
+            worker_id = (worker_info.id + self.worker_offset) % worker_info.num_workers
+            indices = chunks[worker_id]
+        else:
+            worker_id = 0
+            indices = self.indices
+
+        to_skip = (
+            self.to_skip if isinstance(self.to_skip, int) else self.to_skip[worker_id]
+        )
+
+        def documents():
+            for idx in indices:
+                ln = self.len_cache[idx]
+                if ln == 0:
+                    r = self.dataset[idx]
+                    ln = r.shape[0]
+                    self.len_cache[idx] = ln
+                    yield (ln, Cell(r))
+                else:
+                    create = (lambda i: lambda: self.dataset[i])(idx)
```

Comment:
I'm assuming this is the crux of [previous comment to combine into single dataset object](https://github.com/facebookresearch/metaseq/pull/337/files#r976832368), but having a hard time following the indexing and nested lambdas. 😅   

Previously, `DeferredDataset` returned `DeferredTensor(ln, lambda: self.dataset[idx])` for `__getitem__`, but now.... we're indexing into different datasets?

Reply:
This is what is left of DeferredTensor after optimizing the Python by removing custom objects. 

`create` is a zero-argument function that actually performs the load of the tokens from a file. Python variable scoping rules require another function to be introduced to capture the _current_ value of `idx` int `i`, because`idx` will get updated to a new value along the next iteration of the loop.

Since Python has no compiler, optimized code ends up looking harder to understand. In particular, these lambdas/tuples represent the same thing that DeferredTensors did before but without the self-documenting nature of objects. I can add some comments to clarify.

Reply:
I ended up reorganizing and adding some type annotations to remove the lambdas. Should be clearer now.